### PR TITLE
OCPQE-22641: refactor OCP-66839 due to new changes on caps annotations

### PR DESCRIPTION
**Pull request by jiajliu**
1. In 4.15 ipi installation, one more optional operator CloudCredential added, so update expected caps in extracted manifests
2. From 4.15, there is new caps annotation format, such as `MachineAPI+CloudCredential`, so add EXPECTED_CAPABILITIES_IN_MANIFEST to replace original ADDITIONAL_ENABLED_CAPABILITIES.
3. From the same reason with 2, the caps in extracted CR maybe different with the caps in ADDITIONAL_ENABLED_CAPABILITIES, so add EXPECTED_CAPABILITIES_IN_CREDENTIALREQUEST_PRE.
4. Update EXPECTED_CAPABILITIES_IN_CREDENTIALREQUEST to EXPECTED_CAPABILITIES_IN_CREDENTIALREQUEST_POST to be more clear.

### CODE REVIEW

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
cc @jianlinliu for a review, no luck on `exec /tools/entrypoint: argument list too long` issue although trying to drop some steps. i guess we could get it merge without a test since the change is only for ota case, wdyt?

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jianlinliu** [comment]
> cc @jianlinliu for a review, no luck on `exec /tools/entrypoint: argument list too long` issue although trying to drop some steps. i guess we could get it merge without a test since the change is only for ota case, wdyt?

If the change is a bit big, not enough confident, maybe we can replace the test chain with `openshift-upgrade-qe-sanity-disconnected` (of course, maybe we need to add upgrade-precheck and upgrade-postcheck) as a workaround for a luck.

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
> maybe we can replace the test chain with openshift-upgrade-qe-sanity-disconnected (of course, maybe we need to add ota-preupgrade and ota-postupgade) as a workaround for a luck.

@jianlinliu I'm lucky this time, thanks for the suggestion. BTW, could u help review in advance, because this test job may take a long time.

**Comment by jianlinliu** [comment]
The PR looks good to me for now.

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28

**Comment by jiajliu** [comment]
no luck for a successful installation due to quota issue `Acquiring leases for test`, will leave the pr's test after holiday, anyway I ran the case with flexy job to move on.

**Comment by jianlinliu** [comment]
> no luck for a successful installation due to quota issue `Acquiring leases for test`, will leave the pr's test after holiday, anyway I ran the case with flexy job to move on.

@liangxia for now, we still often hit slice quota issue, shall we continue to lower down the frequency of 4.17 ci jbs.

**Comment by liangxia** [comment]
> > no luck for a successful installation due to quota issue `Acquiring leases for test`, will leave the pr's test after holiday, anyway I ran the case with flexy job to move on.
> 
> @liangxia for now, we still often hit slice quota issue, shall we continue to lower down the frequency of 4.17 ci jbs.

Let's give a little more time for the existing destructive jobs to fade away, and check back again next week.

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
@jianlinliu @yunjiang29 any suggestion for the failures [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800687137268436992) and [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28/1800687111494438912)

**Comment by jianlinliu** [comment]
> @jianlinliu @yunjiang29 any suggestion for the failures [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800687137268436992) and [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28/1800687111494438912)

Have you ever hit such issues in your previous pj-rehearse jobs? And I checked a history [job](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/logs/periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800102015724949504) triggered one day ago, did not reproduce this issue.

**Comment by jiajliu** [comment]
> Have you ever hit such issues in your previous pj-rehearse jobs? And I checked a history [job](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/logs/periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800102015724949504) triggered one day ago, did not reproduce this issue.

Nope, it happened from yestoday [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800460483128463360) and [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28/1800460223261970432). I re-tried this morning, still hit the same issue.

**Comment by jianlinliu** [comment]
> > Have you ever hit such issues in your previous pj-rehearse jobs? And I checked a history [job](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/logs/periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800102015724949504) triggered one day ago, did not reproduce this issue.
> 
> Nope, it happened from yestoday [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800460483128463360) and [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28/1800460223261970432). I re-tried this morning, still hit the same issue.

@liangxia was there any update on `verification-tests` image yesterday ?

**Comment by liangxia** [comment]
> > > Have you ever hit such issues in your previous pj-rehearse jobs? And I checked a history [job](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/logs/periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800102015724949504) triggered one day ago, did not reproduce this issue.
> > 
> > 
> > Nope, it happened from yestoday [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1800460483128463360) and [link](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28/1800460223261970432). I re-tried this morning, still hit the same issue.
> 
> @liangxia was there any update on `verification-tests` image yesterday ?

Yes, more detail in https://issues.redhat.com/browse/OCPQE-22684

**Comment by jianlinliu** [comment]
> Nope, it happened from yestoday

I feel like this issue came from `verification-tests` rebuild. Actually I have no idea why `oc adm release info` was trying to access `/opt/app-root/src/.config/containers/registries.conf.d`

@jiajliu can you help add the following code at the beginning of  `cucushift-upgrade-mirror-images-by-oc-adm-commands.sh` for a luck:
```
set -x 

export HOME="${HOME:-/tmp/home}"
export XDG_RUNTIME_DIR="${HOME}/run"
export REGISTRY_AUTH_PREFERENCE=podman # TODO: remove later, used for migrating oc from docker to podman
mkdir -p "${XDG_RUNTIME_DIR}"
```

**Comment by liangxia** [comment]
> > Nope, it happened from yestoday
> 
> I feel like this issue came from `verification-tests` rebuild. Actually I have no idea why `oc adm release info` was trying to access `/opt/app-root/src/.config/containers/registries.conf.d`
> 
> @jiajliu can you help add the following code at the beginning of `cucushift-upgrade-mirror-images-by-oc-adm-commands.sh` for a luck:
> 
> ```
> set -x 
> 
> export HOME="${HOME:-/tmp/home}"
> export XDG_RUNTIME_DIR="${HOME}/run"
> export REGISTRY_AUTH_PREFERENCE=podman # TODO: remove later, used for migrating oc from docker to podman
> mkdir -p "${XDG_RUNTIME_DIR}"
> ```

Besides the rebuild of verification-tests image, it also switch base image from centos:stream8 to rhel-8-base-nodejs-openshift-4.12 as centos:stream8 reach EOL as of May 31

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jianlinliu** [comment]
Okay, failed again. I am going to create a separate PR to debug.

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28

**Comment by jiajliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28

**Comment by jiajliu** [comment]
hi @jianlinliu test done, i just revert the changes for the test and debug. PTAL again, thanks.

4.14-4.15 upgrade test pass in https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-nightly-4.15-upgrade-from-stable-4.14-aws-ipi-disc-priv-sts-basecap-none-f28/1801175851547496448/artifacts/aws-ipi-disc-priv-sts-basecap-none-f28/ , the failure in preupgrade is a known issue https://issues.redhat.com/browse/OCPBUGS-26523

4.15-4.16 upgrade test pass in https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/52862/rehearse-52862-periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-4.16-upgrade-from-stable-4.15-aws-ipi-disc-priv-sts-basecap-none-fips-f28/1801145630752509952/artifacts/aws-ipi-disc-priv-sts-basecap-none-fips-f28/, the failure in preupgrade is also a known issue https://issues.redhat.com/browse/OCPBUGS-35000

File: ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh [hunk 1]
```diff
@@ -107,7 +107,7 @@ function pre-OCP-66839(){
 
     # There should be only enabled cap annotaion in all extracted manifests
     curCap=$(grep -rh "capability.openshift.io/name:" "${manifestsDir}"|awk -F": " '{print $NF}'|sort -u|xargs)
-    expectedCap=$(echo ${ADDITIONAL_ENABLED_CAPABILITIES} | sort -u|xargs)
+    expectedCap=$(echo ${EXPECTED_CAPABILITIES_IN_MANIFEST} | tr ' ' '\n'|sort -u| tr '\n' ' '|xargs)
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:110]:
```suggestion
    expectedCap=$(echo ${EXPECTED_CAPABILITIES_IN_MANIFEST} | tr ' ' '\n'|sort -u|xargs)
```
I guess we do not need the 2nd `tr` here.

Comment (jiajliu) [reply ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh]: updated

**Pull request merged**